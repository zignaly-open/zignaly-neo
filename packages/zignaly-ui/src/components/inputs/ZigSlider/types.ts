import { SliderProps } from "@mui/material";

export type SliderLabels = {
  top?: string | JSX.Element;
  start?: string;
  end?: string;
  className?: string;
  /** Format slider value, by default it will shows value + % if max is 100. Set to null to disable. */
  valueLabelFormat?: (value: number) => string;
  showValues?: boolean;
  /**
   * Left value: max - value, Right value: value
   * Default, Left value: min, Right value: max
   */
  invertSliderValues?: boolean;
};
export type ZigSliderProps = SliderProps & {
  prefixId?: string;
  labels?: SliderLabels;
};
