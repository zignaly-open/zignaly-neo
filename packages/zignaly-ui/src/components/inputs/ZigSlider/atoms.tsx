import { Box } from "@mui/material";
import { ZigTypography } from "index";
import React from "react";
import { SliderLabels } from "./types";

export const SliderLabelValue = ({
  labels,
  side,
  max = 100,
  value = 0,
}: {
  labels: SliderLabels;
  side: "start" | "end";
  max?: number;
  value?: number;
}) => (
  <Box display="flex" flexDirection="column" alignItems="center">
    <ZigTypography color="neutral100" variant="body2">
      {side === "start" ? labels.start : labels.end}
    </ZigTypography>
    <ZigTypography color="neutral400" variant="body2">
      {side === "start"
        ? labels.displayDifference !== false
          ? max - value
          : value
        : labels.displayDifference !== false
        ? value
        : max}
    </ZigTypography>
  </Box>
);
