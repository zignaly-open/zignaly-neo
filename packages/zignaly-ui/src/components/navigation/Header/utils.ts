// TODO: Move this for config or theme folder.
export const headerConfig = {
  height: "52px",
};
