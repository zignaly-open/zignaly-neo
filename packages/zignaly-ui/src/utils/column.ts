import styled from "styled-components";

export const Column = styled.div<{
  gap?: number;
  justifyContent?: "center" | "end" | "start" | "space-between";
  alignItems?: "center" | "end" | "start" | "space-between";
  textAlign?: "end" | "center" | "start";
  height?: string;
  flex?: number;
  width?: string;
}>`
  display: flex;
  flex-direction: column;
  flex: ${(props) => props.flex};
  gap: ${(props) => props.gap}px;
  justify-content: ${(props) => props.justifyContent};
  align-items: ${(props) => props.alignItems};
  text-align: ${(props) => `${props.textAlign}`};
  height: ${(props) => `${props.height}`};
  width: ${(props) => `${props.width}`};
`;
