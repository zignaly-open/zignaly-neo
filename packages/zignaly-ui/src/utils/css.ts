import { css } from "styled-components";

export const NiceScrollbar = css`
  &::-webkit-scrollbar {
    width: 6px;
  }

  /* Track */
  &::-webkit-scrollbar-track {
    background: rgba(255, 255, 255, 0.05);
    border-radius: 12px;
  }

  /* Handle */
  &::-webkit-scrollbar-thumb {
    background: rgba(255, 255, 255, 0.1);
    border-radius: 12px;
  }

  /* Handle on hover */
  &::-webkit-scrollbar-thumb:hover {
    background: rgba(255, 255, 255, 0.15);
  }
`;
