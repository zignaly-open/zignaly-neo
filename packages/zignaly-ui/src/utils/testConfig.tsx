import { render } from "@testing-library/react";
import { ThemeProvider as ThemeProviderMui } from "@mui/material";
import { ThemeProvider, ThemeProviderMui as ThemeInheritorMui } from "../index";
import React from "react";
import { getZignalyThemeExport } from "../theme";
import mediaQuery from "css-mediaquery";
import { JestI18nextDecorator } from "./dummyI18nextProvider";

export const renderWithProvidersUi = (ui: JSX.Element, renderOptions = {}) => {
  const { mui, legacyStyledComponentsDoNotUse } = getZignalyThemeExport();
  return render(ui, {
    wrapper: ({ children }: { children: JSX.Element }) => (
      <JestI18nextDecorator>
        <ThemeProvider theme={legacyStyledComponentsDoNotUse}>
          <ThemeInheritorMui theme={mui}>
            <ThemeProviderMui theme={mui}>{children}</ThemeProviderMui>
          </ThemeInheritorMui>
        </ThemeProvider>
      </JestI18nextDecorator>
    ),
    ...renderOptions,
  });
};

const createMatchMedia = (width: number) => {
  return (query: string) => {
    return {
      matches: mediaQuery.match(query, { width }),
      media: "",
      addListener: () => {},
      removeListener: () => {},
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      dispatchEvent: () => true,
    };
  };
};

export const resizeScreenSize = (width: number) => {
  window.matchMedia = createMatchMedia(width);
};
