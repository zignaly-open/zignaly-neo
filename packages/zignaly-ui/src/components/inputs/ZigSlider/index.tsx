import React, { forwardRef } from "react";
import { Box, Slider } from "@mui/material";
import { ZigSliderProps } from "./types";
import { SliderLabelValue } from "./atoms";
import ZigTypography from "components/display/ZigTypography";

const ZigSlider = forwardRef(
  (
    {
      labels = {},
      prefixId,
      className = "",
      valueLabelFormat,
      labelFormat = true,
      ...props
    }: ZigSliderProps,
    ref,
  ) => {
    let valueLabelFormatDefaulted = valueLabelFormat;
    if (!valueLabelFormatDefaulted && (!props.max || props.max === 100)) {
      valueLabelFormatDefaulted = (value: number) => `${value}%`;
    }
    let labelFormatDefaulted = valueLabelFormatDefaulted;
    if (labelFormat && labelFormat !== true) {
      labelFormatDefaulted = labelFormat;
    }
    const { start, end, showValues = true, labelsAbove } = labels;
    const showLabels = start || end;

    const marks = [25, 50, 75].map((percentage) => ({
      value: (percentage * (props.max || 100)) / 100,
    }));

    return (
      <Box
        display="flex"
        flexDirection="column"
        id={prefixId && `${prefixId}__container`}
        gap="10px"
        width={1}
        className={className}
        ref={ref}
      >
        {labels?.top && (
          <ZigTypography color={"neutral200"} id={prefixId && `${prefixId}__title`}>
            {labels?.top}
          </ZigTypography>
        )}
        <Box
          display="flex"
          width={1}
          gap={1}
          alignItems="center"
          pt={showLabels && labelsAbove !== false ? 3 : 0}
          pb={showLabels && labelsAbove === false ? 3 : 0}
          position="relative"
        >
          <Box display="flex" width={1} gap={"13px"} alignItems="center">
            {showValues && (
              <SliderLabelValue
                prefixId={prefixId && `${prefixId}__label-start`}
                side="start"
                labels={labels}
                min={props.min}
                max={props.max}
                value={props.value as number}
                labelFormat={labelFormatDefaulted}
              />
            )}
            <Slider
              id={prefixId && `${prefixId}__slider`}
              marks={marks}
              track={false}
              valueLabelDisplay={labels.invertSliderValues ? "off" : "auto"}
              valueLabelFormat={valueLabelFormatDefaulted}
              sx={{ width: "auto", flex: 1 }}
              {...props}
            />
            {showValues && (
              <SliderLabelValue
                prefixId={prefixId && `${prefixId}__label-end`}
                side="end"
                labels={labels}
                min={props.min}
                max={props.max}
                value={props.value as number}
                labelFormat={labelFormatDefaulted}
              />
            )}
          </Box>
        </Box>
      </Box>
    );
  },
);

export default ZigSlider;
