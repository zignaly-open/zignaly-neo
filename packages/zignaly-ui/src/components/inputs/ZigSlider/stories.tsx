import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import ZigSlider from "./index";
import { useState } from "react";
import { ZigSliderProps } from "./types";

const meta = {
  title: "Inputs/ZigSlider",
  component: ZigSlider,
} as Meta;
export default meta;

const ZigSliderControlled = (props: ZigSliderProps) => {
  const [value, setValue] = useState(props.value);

  return <ZigSlider {...props} value={value} onChange={(e, v) => setValue(v)} />;
};

type Story = StoryObj<typeof meta>;

export const WithLabel: Story = {
  args: {
    labels: { top: "My Slider", start: "Start", end: "End", invertSliderValues: true },
  },
  render: (props) => <ZigSliderControlled {...props} />,
};

export const WithoutLabels: Story = { args: {} };
