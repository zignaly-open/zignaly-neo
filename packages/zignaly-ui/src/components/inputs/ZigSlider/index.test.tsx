import React from "react";
import { renderWithProvidersUi } from "../../../utils/testConfig";
import { fireEvent, screen } from "@testing-library/react";
import ZigSlider from ".";

describe("components/inputs/ZigSlider", () => {
  const defaultProps = {
    min: 0,
    max: 100,
    value: 50,
    labels: {
      start: "Start Label",
      end: "End Label",
      top: "Top Label",
    },
  };

  it("should render without crashing", () => {
    const { container } = renderWithProvidersUi(<ZigSlider {...defaultProps} prefixId={"test"} />);

    const sliderContainer = container.querySelector("#test__container") as HTMLElement;
    expect(sliderContainer).toBeVisible();
    expect(sliderContainer).toMatchSnapshot();
  });

  it("should display labels", () => {
    renderWithProvidersUi(<ZigSlider {...defaultProps} />);

    const startLabel = screen.getByText("Start Label");
    const endLabel = screen.getByText("End Label");
    const topLabel = screen.getByText("Top Label");

    expect(startLabel).toBeVisible();
    expect(endLabel).toBeVisible();
    expect(topLabel).toBeVisible();
  });

  it("should handle slider value change", () => {
    const onChangeMock = jest.fn();
    renderWithProvidersUi(<ZigSlider {...defaultProps} onChange={onChangeMock} />);

    const slider = screen.getByRole("slider");
    fireEvent.change(slider, { target: { value: 75 } });

    expect(onChangeMock).toHaveBeenCalledWith(expect.anything(), 75, expect.anything());
  });

  it("should display custom formatted labels", () => {
    const customFormat = (value: number) => `Formatted: ${value}%`;
    renderWithProvidersUi(<ZigSlider {...defaultProps} valueLabelFormat={customFormat} />);

    const startLabel = screen.getByText("Formatted: 0%");
    const endLabel = screen.getByText("Formatted: 100%");

    expect(startLabel).toBeVisible();
    expect(endLabel).toBeVisible();
  });
});
