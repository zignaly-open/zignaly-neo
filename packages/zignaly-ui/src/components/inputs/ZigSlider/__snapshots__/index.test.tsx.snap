// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/inputs/ZigSlider should render without crashing 1`] = `
<div
  class="MuiBox-root css-ntktx3"
  id="test__container"
>
  <span
    class="MuiTypography-root MuiTypography-body1 css-a54hgx-MuiTypography-root"
    id="test__title"
  >
    Top Label
  </span>
  <div
    class="MuiBox-root css-gn04vd"
  >
    <div
      class="MuiBox-root css-1adenbn"
    >
      <div
        class="MuiBox-root css-1hnllux"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 css-u6htvk-MuiTypography-root"
          id="test__label-start-label"
        >
          Start Label
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 css-xsotb5-MuiTypography-root"
          id="test__label-start-value"
        >
          0%
        </span>
      </div>
      <span
        class="MuiSlider-root MuiSlider-trackFalse MuiSlider-colorPrimary MuiSlider-sizeMedium css-rmijdq-MuiSlider-root"
        id="test__slider"
      >
        <span
          class="MuiSlider-rail css-47oe14-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-1cwtxq1-MuiSlider-track"
          style="left: 0%; width: 50%;"
        />
        <span
          class="MuiSlider-mark css-vjt6xi-MuiSlider-mark"
          data-index="0"
          style="left: 25%;"
        />
        <span
          class="MuiSlider-mark MuiSlider-markActive css-nmrk1k-MuiSlider-mark"
          data-index="1"
          style="left: 50%;"
        />
        <span
          class="MuiSlider-mark css-vjt6xi-MuiSlider-mark"
          data-index="2"
          style="left: 75%;"
        />
        <span
          class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-1ccmxfh-MuiSlider-thumb"
          data-index="0"
          style="left: 50%;"
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="50"
            data-index="0"
            max="100"
            min="0"
            step="1"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="50"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-1v43nqo-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                50%
              </span>
            </span>
          </span>
        </span>
      </span>
      <div
        class="MuiBox-root css-1h56rwg"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 css-1ctd5td-MuiTypography-root"
          id="test__label-end-label"
        >
          End Label
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 css-xsotb5-MuiTypography-root"
          id="test__label-end-value"
        >
          100%
        </span>
      </div>
    </div>
  </div>
</div>
`;
